optics/src/Optics.hs

```diff
@@ -224,52 +199,356 @@ import Data.Either.Optics                    as P
 -- generates a type error if the composition does not make sense.
+--
+-- The optic kind resulting from a composition is the least upper bound (join)
+-- of the optic kinds being composed, if it exists.  The 'Join' type family
+-- computes the least upper bound given two optic kind tags.  For example the
+-- 'Join' of a 'Lens' and a 'Prism' is an 'AffineTraversal'.
+--
+-- >>> :kind! Join A_Lens A_Prism
+-- Join A_Lens A_Prism :: *
+-- = An_AffineTraversal
+--
+-- The join does not exist for some pairs of optic kinds, which means that they
+-- cannot be composed.  For example there is no optic kind above both 'Setter'
+-- and 'Fold':
+--
+-- >>> :kind! Join A_Setter A_Fold
+-- Join A_Setter A_Fold :: *
+-- = (TypeError ...)
+--
+-- >>> :t mapped % folded
+-- ...
+-- ...A_Setter cannot be composed with A_Fold
+-- ...
 
-
--- $advantages
+-- $lens_comparison
+--
+-- The @lens@ package is the best known Haskell library for optics, and
+-- established many of the foundations on which the @optics@ package builds (not
+-- least in quite a bit of code having been directly ported).  It defines optics
+-- based on the /van Laarhoven/ representation, where each optic kind is
+-- introduced as a /transparent/ type synonym for a complex polymorphic type,
+-- for example:
+--
+-- @
+-- type Lens s t a b = forall f. 'Functor' f => (a -> f b) -> s -> f t
+-- @
+--
+-- In contrast, @optics@ tries to preserve an abstraction boundary between the
+-- interface of optics and their implementation.  Optic kinds are expressed
+-- directly in the types, as 'Optic' is an /opaque/ newtype:
+--
+-- @
+-- type 'Lens' s t a b = 'Optic' 'A_Lens' 'NoIx' s t a b
+-- @
+--
+-- The choice of representation of 'Optic' is then an implementation detail, not
+-- essential for understanding the library.  (In fact, @optics@ uses the
+-- /profunctor/ representation rather than the /van Laarhoven/ representation;
+-- this affects the optic kinds and operations that can be conveniently
+-- supported, but not the essence of the design.)
 --
--- In general, this leads to better results from type inference (the optic kind
--- is preserved in the inferred type):
+-- Our design choice to use /opaque/ rather than /transparent/ abstractions
+-- leads to various consequences, both positive and negative, which are explored
+-- in the following subsections.
+--
+-- == Advantages of the opaque design
+--
+-- Since the interface is deliberately chosen rather than to some extent
+-- determined by the implementation, we are free to choose a more restricted
+-- interface where doing so leads to conceptual simplicity.  For example, in
+-- @lens@, the 'view' function can be used with a 'Fold' provided the result
+-- type has a 'Monoid' instance, and the multiple targets of the 'Fold' will be
+-- combined monoidally.  This behaviour can be confusing, so in @optics@ a
+-- 'Getter' cannot be silently used as a 'Fold', and we prefer to have 'view'
```

Comment:
I think you meant the other way around, i.e. "a Fold cannot be silently used as a Getter".

Reply:
Well spotted! Fixed.